In [1]:
using DataFrames, CSV, JuMP, Gurobi, Plots, Random

InterruptException: InterruptException:

In [2]:
const GRB_ENV = Gurobi.Env(output_flag=0);

In [3]:
function team_match()
    # create model
    model = Model(() -> Gurobi.Optimizer(GRB_ENV))
    set_optimizer_attribute(model, "TimeLimit", 300)

    # PARAMETERS
    N = 20 # teams
    W = 38

    # VARIABLES
    @variable(model, x[i = 1:N, j = 1:N, w = 1:W], Bin) # 1 if team i plays team j on week w, 0 otherwise

    # OBJECTIVE FUNCTION
    @objective(model, Max, sum(x[i,j,w] for i in 1:N, j in 1:N, w in 1:W))

    # CONSTRAINTS
    # teams can never play against themselves
    @constraint(model, [i = 1:N, w = 1:W], x[i, i, w] == 0)

    # each team can only play at most one game per week
    @constraint(model, [i = 1:N, w = 1:W], sum(x[i, j, w] + x[j, i, w] for j in 1:N) <= 1)

    for i=1:N
        for j=1:N
            if i!=j
                # each game is against a different team, team play against each other team at most once in first half of season
                @constraint(model, sum(x[i, j, w] + x[j, i, w] for w in 1:Int(W/2)) == 1)
                
                # each game is against a different team, team play against each other team at most once in second half of season
                @constraint(model, sum(x[i, j, w] + x[j, i, w] for w in Int((W/2))+1:W) == 1)

                # every team plays every other team once at home
                @constraint(model, sum(x[i, j, w] for w in 1:W) == 1)
                
            end
        end
    end

    # no more than 2 straight home games
    @constraint(model, [i = 1:N, w = 1:W-2], sum(x[i, :, w] + x[i, :, w+1] + x[i, :, w+2]) <= 2)

    # no more than 2 straight away games
    @constraint(model, [j = 1:N, w = 1:W-2], sum(x[:, j, w] + x[:, j, w+1] + x[:, j, w+2]) <= 2)


    # OPTIMIZE
    # solvetime = @elapsed optimize!(model)
    optimize!(model)


    return value.(x)
end

team_match (generic function with 1 method)

In [4]:
x = team_match()

20×20×38 Array{Float64, 3}:
[:, :, 1] =
  0.0  -0.0  -0.0  -0.0  -0.0   1.0  …   0.0  -0.0   0.0  -0.0   0.0  -0.0
 -0.0   0.0  -0.0  -0.0  -0.0  -0.0     -0.0  -0.0  -0.0   0.0  -0.0  -0.0
  0.0  -0.0   0.0  -0.0  -0.0  -0.0     -0.0  -0.0  -0.0  -0.0  -0.0  -0.0
 -0.0   0.0  -0.0   0.0  -0.0  -0.0     -0.0  -0.0  -0.0  -0.0  -0.0  -0.0
  0.0  -0.0  -0.0  -0.0   0.0  -0.0     -0.0  -0.0   0.0  -0.0  -0.0  -0.0
 -0.0  -0.0  -0.0  -0.0  -0.0   0.0  …  -0.0  -0.0  -0.0  -0.0  -0.0  -0.0
 -0.0  -0.0  -0.0  -0.0  -0.0  -0.0     -0.0  -0.0  -0.0  -0.0  -0.0  -0.0
 -0.0  -0.0   1.0  -0.0  -0.0  -0.0     -0.0  -0.0  -0.0  -0.0   0.0  -0.0
 -0.0  -0.0  -0.0  -0.0  -0.0  -0.0     -0.0  -0.0  -0.0  -0.0  -0.0  -0.0
 -0.0   1.0  -0.0  -0.0  -0.0  -0.0     -0.0  -0.0  -0.0  -0.0  -0.0  -0.0
 -0.0  -0.0  -0.0  -0.0   1.0  -0.0  …  -0.0  -0.0  -0.0  -0.0  -0.0  -0.0
 -0.0  -0.0  -0.0   1.0  -0.0  -0.0     -0.0  -0.0  -0.0  -0.0  -0.0  -0.0
 -0.0  -0.0  -0.0  -0.0  -0.0  -0.0      0.0  -0.0   0.0  -0

In [ ]:
function table_games_w(x, team)
    ## create w dataframes, one for each week, each dataframe contains the games for that week
    N = 20
    W = 38
    table = zeros(W,2)
    for w=1:W
        for j=1:N
            if x[team,j,w] == 1 && x[j,team,w] == 0
                table[w,1] = team
                table[w,2] = j 
            elseif x[team,j,w] == 0 && x[j,team,w] == 1
                table[w,1] = j
                table[w,2] = team
            end
        end
    end
    return table
end

table_games_w (generic function with 1 method)

In [ ]:
function all_games(x)
     
    N = 20
    W = 38

    team1 = table_games_w(x,1)
    team2 = table_games_w(x,2)
    team3 = table_games_w(x,3)
    team4 = table_games_w(x,4)
    team5 = table_games_w(x,5)
    team6 = table_games_w(x,6)
    team7 = table_games_w(x,7)
    team8 = table_games_w(x,8)
    team9 = table_games_w(x,9)
    team10 = table_games_w(x,10)
    team11 = table_games_w(x,11)
    team12 = table_games_w(x,12)
    team13 = table_games_w(x,13)
    team14 = table_games_w(x,14)
    team15 = table_games_w(x,15)
    team16 = table_games_w(x,16)
    team17 = table_games_w(x,17)
    team18 = table_games_w(x,18)
    team19 = table_games_w(x,19)
    team20 = table_games_w(x,20)

    table = zeros(10,W*2)
    for w=1:W
        c = 1
        if team1[w,1] == 1
            table[c,w*2-1] = team1[w,1]
            table[c,w*2] = team1[w,2]
            c += 1
        end 
        if team2[w,1] == 2
            table[c,w*2-1] = team2[w,1]
            table[c,w*2] = team2[w,2]
            c += 1
        end
        if team3[w,1] == 3
            table[c,w*2-1] = team3[w,1]
            table[c,w*2] = team3[w,2]
            c += 1
        end
        if team4[w,1] == 4
            table[c,w*2-1] = team4[w,1]
            table[c,w*2] = team4[w,2]
            c += 1
        end
        if team5[w,1] == 5
            table[c,w*2-1] = team5[w,1]
            table[c,w*2] = team5[w,2]
            c += 1
        end
        if team6[w,1] == 6
            table[c,w*2-1] = team6[w,1]
            table[c,w*2] = team6[w,2]
            c += 1
        end
        if team7[w,1] == 7
            table[c,w*2-1] = team7[w,1]
            table[c,w*2] = team7[w,2]
            c += 1
        end
        if team8[w,1] == 8
            table[c,w*2-1] = team8[w,1]
            table[c,w*2] = team8[w,2]
            c += 1
        end
        if team9[w,1] == 9
            table[c,w*2-1] = team9[w,1]
            table[c,w*2] = team9[w,2]
            c += 1
        end
        if team10[w,1] == 10
            table[c,w*2-1] = team10[w,1]
            table[c,w*2] = team10[w,2]
            c += 1
        end
        if team11[w,1] == 11
            table[c,w*2-1] = team11[w,1]
            table[c,w*2] = team11[w,2]
            c += 1
        end
        if team12[w,1] == 12
            table[c,w*2-1] = team12[w,1]
            table[c,w*2] = team12[w,2]
            c += 1
        end
        if team13[w,1] == 13
            table[c,w*2-1] = team13[w,1]
            table[c,w*2] = team13[w,2]
            c += 1
        end
        if team14[w,1] == 14
            table[c,w*2-1] = team14[w,1]
            table[c,w*2] = team14[w,2]
            c += 1
        end
        if team15[w,1] == 15
            table[c,w*2-1] = team15[w,1]
            table[c,w*2] = team15[w,2]
            c += 1
        end
        if team16[w,1] == 16
            table[c,w*2-1] = team16[w,1]
            table[c,w*2] = team16[w,2]
            c += 1
        end
        if team17[w,1] == 17
            table[c,w*2-1] = team17[w,1]
            table[c,w*2] = team17[w,2]
            c += 1
        end
        if team18[w,1] == 18
            table[c,w*2-1] = team18[w,1]
            table[c,w*2] = team18[w,2]
            c += 1
        end
        if team19[w,1] == 19
            table[c,w*2-1] = team19[w,1]
            table[c,w*2] = team19[w,2]
            c += 1
        end
        if team20[w,1] == 20
            table[c,w*2-1] = team20[w,1]
            table[c,w*2] = team20[w,2]
            c += 1
        end
    end

    ### change table into dataframe
    df_games = DataFrame(W1_Home=table[:,1], W1_Away=table[:,2], W2_Home=table[:,3], W2_Away=table[:,4], W3_Home=table[:,5], W3_Away=table[:,6], W4_Home=table[:,7], W4_Away=table[:,8], W5_Home=table[:,9], W5_Away=table[:,10], W6_Home=table[:,11], W6_Away=table[:,12], W7_Home=table[:,13], W7_Away=table[:,14], W8_Home=table[:,15], W8_Away=table[:,16], W9_Home=table[:,17], W9_Away=table[:,18], W10_Home=table[:,19], W10_Away=table[:,20], W11_Home=table[:,21], W11_Away=table[:,22], W12_Home=table[:,23], W12_Away=table[:,24], W13_Home=table[:,25], W13_Away=table[:,26], W14_Home=table[:,27], W14_Away=table[:,28], W15_Home=table[:,29], W15_Away=table[:,30], W16_Home=table[:,31], W16_Away=table[:,32], W17_Home=table[:,33], W17_Away=table[:,34], W18_Home=table[:,35], W18_Away=table[:,36], W19_Home=table[:,37], W19_Away=table[:,38], W20_Home=table[:,39], W20_Away=table[:,40], W21_Home=table[:,41], W21_Away=table[:,42], W22_Home=table[:,43], W22_Away=table[:,44], W23_Home=table[:,45], W23_Away=table[:,46], W24_Home=table[:,47], W24_Away=table[:,48], W25_Home=table[:,49], W25_Away=table[:,50], W26_Home=table[:,51], W26_Away=table[:,52], W27_Home=table[:,53], W27_Away=table[:,54], W28_Home=table[:,55], W28_Away=table[:,56], W29_Home=table[:,57], W29_Away=table[:,58], W30_Home=table[:,59], W30_Away=table[:,60], W31_Home=table[:,61], W31_Away=table[:,62], W32_Home=table[:,63], W32_Away=table[:,64], W33_Home=table[:,65], W33_Away=table[:,66], W34_Home=table[:,67], W34_Away=table[:,68], W35_Home=table[:,69], W35_Away=table[:,70], W36_Home=table[:,71], W36_Away=table[:,72], W37_Home=table[:,73], W37_Away=table[:,74], W38_Home=table[:,75], W38_Away=table[:,76])

    return df_games
end         

all_games (generic function with 1 method)

In [ ]:
all_games(x)

10×76 DataFrame
 Row │ W1_Home  W1_Away  W2_Home  W2_Away  W3_Home  W3_Away  W4_Home  W4_Away  ⋯
     │ Float64  Float64  Float64  Float64  Float64  Float64  Float64  Float64  ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │     2.0     12.0      1.0      2.0      2.0     17.0      1.0     12.0  ⋯
   2 │     4.0     19.0      3.0      6.0      4.0      1.0      2.0     11.0
   3 │     5.0      1.0      8.0     20.0      5.0     18.0      5.0     13.0
   4 │     7.0      3.0      9.0     10.0      6.0      9.0      6.0     10.0
   5 │     8.0     18.0     11.0     12.0     10.0     13.0      7.0     20.0  ⋯
   6 │     9.0     20.0     13.0     16.0     11.0      8.0      9.0      8.0
   7 │    10.0     11.0     15.0      4.0     12.0     19.0     14.0      4.0
   8 │    13.0     14.0     17.0      5.0     14.0      7.0     17.0      3.0
   9 │    15.0     17.0     18.0     14.0     16.0     15.0     18.0     16.0  ⋯
  10 │    16.0      6.0     19.0      7.0     20.0      3.0     19.0     15.0
                                                              68 columns omitted

In [5]:
# C is upper triangular matrix; each entry is the cost of playing in t and then t_
function GetC(x,y)
    if x == y
        return 10
    end
    return 1/(y-x)
end

GetC (generic function with 1 method)

In [6]:
# U and Z matrices
### Teams 1,2 play on weeks 5,6,7,12,13,14 on Tuesday
### Teams 3,4 play on weeks 5,6,7,12,13,14 on Wednesday
### Teams 5,6 play on weeks 5,6,7,12,13,14 on Thursday
u = zeros((20,39)) #adding week 39 just so the model can factor in the last week
z = zeros((20,39))
for i in 1:6
    for w in 1:38
        if w == 5 || w == 6 || w == 7 || w == 12 || w == 13 || w ==14 
            u[i,w] = 1.0
            if i == 1 || i == 2
                z[i,w] = 2.0
            elseif i == 3 || i == 4
                z[i,w] = 3.0
            elseif i == 5 || i == 6
                z[i,w] = 4.0
            end
        end
    end
end
z

20×39 Matrix{Float64}:
 0.0  0.0  0.0  0.0  2.0  2.0  2.0  0.0  …  0.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0  2.0  2.0  2.0  0.0     0.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0  3.0  3.0  3.0  0.0     0.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0  3.0  3.0  3.0  0.0     0.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0  4.0  4.0  4.0  0.0     0.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0  4.0  4.0  4.0  0.0  …  0.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0     0.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0     0.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0     0.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0     0.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  …  0.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0     0.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  

In [7]:
function GetC1()
    # C is upper triangular matrix; each entry is the cost of playing in t and then t_
    D=7
    W_ = 39

    C = zeros((D*(W_+1), D*(W_+1)))

    for t in 1:(W_*D)
        for t_ in 1:(W_*D)
            if t < t_
                C[t, t_] = (t_ - t)^(-1)
            end
        end
    end

    return C
end
C = GetC1()

280×280 Matrix{Float64}:
 0.0  1.0  0.5  0.333333  0.25      …  0.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  1.0  0.5       0.333333     0.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  1.0       0.5          0.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0       1.0          0.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0       0.0          0.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0       0.0       …  0.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0       0.0          0.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0       0.0          0.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0       0.0          0.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0       0.0          0.0  0.0  0.0  0.0  0.0  0.0  0.0
 ⋮                                  ⋱            ⋮                   
 0.0  0.0  0.0  0.0       0.0          0.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0       0.0          0.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0       0.0

In [15]:
function schedule_days(x,u,z)
    N=20
    W=38
    D=7 #playable number of days

    # create model
    model = Model(() -> Gurobi.Optimizer(GRB_ENV))
    set_optimizer_attribute(model, "TimeLimit", 21600)
    set_optimizer_attribute(model, "OutputFlag", 1)
    set_optimizer_attribute(model, "MIPGap", 1e-4)

    # VARIABLES
    @variable(model, y[i = 1:N, w = 1:W+1, d = 4:D], Bin) # 1 if team i plays on day d of week w, 0 otherwise
    #filler week W+1 to help with model

    # OBJECTIVE FUNCTION -- rest of one team doesn't depend on the other team, put that in the constraints
    @objective(model, Min, 
        sum( ( (1-u[i,w+1])*GetC(d,7+d_) + u[i,w+1]*(GetC(d,7+z[i,w+1])+GetC(z[i,w+1], d_)) )
           * (y[i,w,d]*y[i,w+1,d_]) 
        for i in 1:N, w in 1:W, d in 4:D, d_ in 4:D))
    # @objective(model, Min, 
    #         sum( ( C[7 * w + d, 7 * (w+1) + d_] )
    #             * (y[i,w,d]*y[i,w+1,d_])
    #         for i in 1:N, w in 1:W, d in 4:D, d_ in 4:D))

    # # OBJECTIVE FUNCTION
    # @objective(model, Min, 
    #     sum( ( (1-u[i,w+1])*GetC(d,7+d_) + u[i,w+1]*(GetC(d,7+z[i,w+1])+GetC(z[i,w+1], d_)) )

    #         * (
    #             (x[i,j,w]+x[j,i,w])*(x[i,k,w+1]+x[k,i,w+1])
    #         * y[i,w,d]*y[i,w+1,d_]
    #         )

    #     for i in 1:N, j in 1:N, k in 1:N, w in 1:W, d in 4:D, d_ in 4:D))

    # CONSTRAINTS
    
    # teams who play each other on week w must play on the same day of the week
    for w=1:W
        for i=1:N
            for j=1:N
                if x[i,j,w]+x[j,i,w] == 1
                    @constraint(model, (y[i,w,:] .== y[j,w,:]))
                end
            end
        end
    end

    # filler week 39
    @constraint(model, [i = 1:N], y[i,W+1,7] == 1)
    
    #team can only play as many times as scheduled
    @constraint(model, [i = 1:N, w = 1:W], sum(y[i,w,:]) == sum(x[i,:,w]+x[:,i,w]))

    # maximum of 1 game on day 1
    @constraint(model, [w = 1:W], sum(y[:,w,4]) <= 2) #2 teams play per game

    # maximum of 4 games on day 2
    @constraint(model, [w = 1:W], sum(y[:,w,5]) <= 10)

    # maximum of 4 games on day 3
    @constraint(model, [w = 1:W], sum(y[:,w,6]) <= 10)

    # maximum of 1 game on day 4
    @constraint(model, [w = 1:W], sum(y[:,w,7]) <= 4)

    # minimum of 3 games on day 2
    @constraint(model, [w = 1:W], sum(y[:,w,5]) >= 6)

    # minimum of 2 games on day 3
    @constraint(model, [w = 1:W], sum(y[:,w,6]) >= 4)

    # at least 15 weekend home games
    @constraint(model, [i = 1:N], sum(y[i,:,5]) + sum(y[i,:,6]) >= 12)

    # # at least 1 Fri home game
    # @constraint(model, [i = 1:N], sum(y[i,:,4]) >= 1)

    # # at least 1 Mon home game
    # @constraint(model, [i = 1:N], sum(y[i,:,7]) >= 1)


    # OPTIMIZE
    # solvetime = @elapsed optimize!(model)
    optimize!(model)


    return value.(y)
end

schedule_days (generic function with 1 method)

In [16]:
y = schedule_days(x,u,z)

Set parameter OutputFlag to value 1
Set parameter OutputFlag to value 1
Gurobi Optimizer version 9.5.2 build v9.5.2rc0 (mac64[arm])
Thread count: 8 physical cores, 8 logical processors, using up to 8 threads
Optimize a model with 4108 rows, 3120 columns and 16820 nonzeros
Model fingerprint: 0x45d78919
Model has 12160 quadratic objective terms
Variable types: 0 continuous, 3120 integer (3120 binary)
Coefficient statistics:
  Matrix range     [1e+00, 1e+00]
  Objective range  [0e+00, 0e+00]
  QObjective range [2e-01, 2e+01]
  Bounds range     [0e+00, 0e+00]
  RHS range        [1e+00, 1e+01]
Presolve removed 3460 rows and 1540 columns
Presolve time: 0.04s
Presolved: 12728 rows, 13660 columns, 42380 nonzeros
Variable types: 0 continuous, 13660 integer (13660 binary)
Found heuristic solution: objective 132.4353175

Root relaxation: objective 2.311111e+00, 2638 iterations, 0.08 seconds (0.11 work units)

    Nodes    |    Current Node    |     Objective Bounds      |     Work
 Expl Unexpl | 

3-dimensional DenseAxisArray{Float64,3,...} with index sets:
    Dimension 1, Base.OneTo(20)
    Dimension 2, Base.OneTo(39)
    Dimension 3, 4:7
And data, a 20×39×4 Array{Float64, 3}:
[:, :, 4] =
 0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  …  0.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0     0.0  0.0  1.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0     0.0  0.0  0.0  0.0  1.0  0.0  0.0
 0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0     0.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0     0.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  …  0.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0     0.0  1.0  0.0  0.0  1.0  0.0  0.0
 0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0     0.0  0.0  0.0  0.0  0.0  0.0  0.0
 1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0     1.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0     0.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0  0.

In [13]:
sum(y[:,1:38,7])

32.0

In [12]:
x[19,7,1]

1.0

In [29]:
function schedule_days_lr(x,u,z)
    N=20
    W=38
    D=7 #playable number of days

    # create model
    model = Model(() -> Gurobi.Optimizer(GRB_ENV))
    set_optimizer_attribute(model, "TimeLimit", 21600)
    set_optimizer_attribute(model, "OutputFlag", 1)
    set_optimizer_attribute(model, "MIPGap", 1e-4)

    # VARIABLES
    @variable(model, y[i = 1:N, w = 1:W+1, d = 4:D], Bin) # 1 if team i plays on day d of week w, 0 otherwise
    @variable(model, abs_f[w = 1:W, z=1:2] >= 0, Int)
    @variable(model, abs_sa[w = 1:W, z=1:2] >= 0, Int)
    @variable(model, abs_su[w = 1:W, z=1:2] >= 0, Int)
    @variable(model, abs_m[w = 1:W, z=1:2] >= 0, Int)
    #@variable(model, P) 
    #@variable(model, lambda[i = 1:4, z=1:2]>=0)
    #filler week W+1 to help with model

    # the bigger the value, the tighter the constraint
    l_1_day = 0.015          #punish from exactness
    l_2_day = 0.01          #punish for breaking range
    l_1_end = 0.005          #punish for exactness
    l_2_end = 0.01          #punish for breaking range

    # OBJECTIVE FUNCTION -- rest of one team doesn't depend on the other team, put that in the constraints
    @objective(model, Min, sum( ((1-u[i,w+1])*GetC(d,7+d_) + u[i,w+1]*(GetC(d,7+z[i,w+1])+GetC(z[i,w+1], d_))) * (y[i,w,d]*y[i,w+1,d_]) for i in 1:N, w in 1:W, d in 4:D, d_ in 4:D)
        + sum(l_1_day*(abs_f[w,1]) + l_1_end*(abs_sa[w,1]) + l_1_end*(abs_su[w,1]) + l_1_day*(abs_m[w,1]) 
        + l_2_day*(abs_f[w,2]) + l_2_end*(abs_sa[w,2]) + l_2_end*(abs_su[w,2]) + l_2_day*(abs_m[w,2]) for w=1:W)
        )

    # @objective(model, Min, 
    #         sum( ( C[7 * w + d, 7 * (w+1) + d_] )
    #             * (y[i,w,d]*y[i,w+1,d_])
    #         for i in 1:N, w in 1:W, d in 4:D, d_ in 4:D))

    # CONSTRAINTS
    # punish schedule for deviating from 1, 4, 4, 1 schedule
    # punish schedule even more for deviating from 0-2, 3-5, 3-5, 0-2 schedule
    @constraint(model, [w = 1:W], abs_f[w,2] >= (sum(y[:,w,4]) - 6))
    @constraint(model, [w = 1:W], abs_f[w,1] >= (sum(y[:,w,4]) - 2))
    @constraint(model, [w = 1:W], abs_f[w,1] >= (2 - sum(y[:,w,4])))

    @constraint(model, [w = 1:W], abs_sa[w,2] >= (sum(y[:,w,5]) - 10))
    @constraint(model, [w = 1:W], abs_sa[w,2] >= (6 - sum(y[:,w,5])))
    @constraint(model, [w = 1:W], abs_sa[w,1] >= (sum(y[:,w,5]) - 8))
    @constraint(model, [w = 1:W], abs_sa[w,1] >= (8 - sum(y[:,w,5])))

    @constraint(model, [w = 1:W], abs_su[w,2] >= (sum(y[:,w,6]) - 10))
    @constraint(model, [w = 1:W], abs_su[w,2] >= (6 - sum(y[:,w,6])))
    @constraint(model, [w = 1:W], abs_su[w,1] >= (sum(y[:,w,6]) - 8))
    @constraint(model, [w = 1:W], abs_su[w,1] >= (8 - sum(y[:,w,6])))

    @constraint(model, [w = 1:W], abs_m[w,2] >= (sum(y[:,w,7]) - 6))
    @constraint(model, [w = 1:W], abs_m[w,1] >= (sum(y[:,w,4]) - 4))
    @constraint(model, [w = 1:W], abs_m[w,1] >= (2 - sum(y[:,w,7])))
   


    # teams who play each other on week w must play on the same day of the week
    for w=1:W
        for i=1:N
            for j=1:N
                if x[i,j,w]+x[j,i,w] == 1
                    @constraint(model, (y[i,w,:] .== y[j,w,:]))
                end
            end
        end
    end

    # @constraint(model, [i = 1:N, j = 1:N, w = 1:W, d = 4:D], y[i,w,d] >= (x[i,j,w]+x[j,i,w])*y[j,w,d])

    # filler week 39
    @constraint(model, [i = 1:N], y[i,W+1,7] == 1)
    
    #team can only play as many times as scheduled
    @constraint(model, [i = 1:N, w = 1:W], sum(y[i,w,:]) == sum(x[i,:,w]+x[:,i,w]))

    # # maximum of 6 game on day 2
    # @constraint(model, [w = 1:W], sum(y[:,w,5]) <= 12) #2 teams play per game

    # # maximum of 6 game on day 3
    # @constraint(model, [w = 1:W], sum(y[:,w,6]) <= 12) #2 teams play per game

    # # minimum of 3 game on day 2
    # @constraint(model, [w = 1:W], sum(y[:,w,5]) >= 6) #2 teams play per game

    # # minimum of 3 game on day 3
    # @constraint(model, [w = 1:W], sum(y[:,w,6]) >= 6) #2 teams play per game

    # at least 12 weekend home games
    @constraint(model, [i = 1:N], sum(y[i,:,5]) + sum(y[i,:,6]) >= 12)


    # OPTIMIZE
    # solvetime = @elapsed optimize!(model)
    optimize!(model)


    return value.(y)
end

schedule_days_lr (generic function with 1 method)

In [263]:
y_lr = schedule_days_lr(x,u,z)

Set parameter OutputFlag to value 1
Set parameter OutputFlag to value 1
Gurobi Optimizer version 9.5.2 build v9.5.2rc0 (mac64[arm])
Thread count: 8 physical cores, 8 logical processors, using up to 8 threads
Optimize a model with 4372 rows, 3424 columns and 21872 nonzeros
Model fingerprint: 0x736ed4ac
Model has 12160 quadratic objective terms
Variable types: 0 continuous, 3424 integer (3120 binary)
Coefficient statistics:
  Matrix range     [1e+00, 1e+00]
  Objective range  [5e-03, 1e-02]
  QObjective range [2e-01, 2e+01]
  Bounds range     [0e+00, 0e+00]
  RHS range        [1e+00, 1e+01]
Found heuristic solution: objective 6.080000e+09
Presolve removed 3440 rows and 1560 columns
Presolve time: 0.04s
Presolved: 12932 rows, 13864 columns, 44932 nonzeros
Found heuristic solution: objective 203.2219048
Variable types: 0 continuous, 13864 integer (13560 binary)

Root relaxation: objective 2.338889e+00, 2368 iterations, 0.10 seconds (0.13 work units)

    Nodes    |    Current Node    |    

3-dimensional DenseAxisArray{Float64,3,...} with index sets:
    Dimension 1, Base.OneTo(20)
    Dimension 2, Base.OneTo(39)
    Dimension 3, 4:7
And data, a 20×39×4 Array{Float64, 3}:
[:, :, 4] =
 1.0  1.0  0.0  1.0  0.0  0.0  0.0  1.0  …  0.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0     0.0  1.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0     0.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0     0.0  0.0  1.0  0.0  0.0  0.0  0.0
 1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0     0.0  0.0  0.0  0.0  0.0  0.0  0.0
 1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  …  0.0  0.0  0.0  0.0  0.0  0.0  0.0
 1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0     0.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0  0.0  1.0  1.0  1.0     0.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0  1.0  1.0  0.0  0.0     0.0  0.0  0.0  0.0  1.0  1.0  0.0
 0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0     0.0  0.0  0.0  0.0  0.0  0.0  0.0
 1.0  1.0  1.0  0.0  0.

Set parameter OutputFlag to value 1
Set parameter OutputFlag to value 1
Gurobi Optimizer version 9.5.2 build v9.5.2rc0 (mac64[arm])
Thread count: 8 physical cores, 8 logical processors, using up to 8 threads
Optimize a model with 4372 rows, 3424 columns and 21872 nonzeros
Model fingerprint: 0x736ed4ac
Model has 12160 quadratic objective terms
Variable types: 0 continuous, 3424 integer (3120 binary)
Coefficient statistics:
  Matrix range     [1e+00, 1e+00]
  Objective range  [5e-03, 1e-02]
  QObjective range [2e-01, 2e+01]
  Bounds range     [0e+00, 0e+00]
  RHS range        [1e+00, 1e+01]
Found heuristic solution: objective 6.080000e+09
Presolve removed 3440 rows and 1560 columns
Presolve time: 0.04s
Presolved: 12932 rows, 13864 columns, 44932 nonzeros
Found heuristic solution: objective 203.2219048
Variable types: 0 continuous, 13864 integer (13560 binary)

Root relaxation: objective 2.338889e+00, 2368 iterations, 0.10 seconds (0.13 work units)

    Nodes    |    Current Node    |     Objective Bounds      |     Work
 Expl Unexpl |  Obj  Depth IntInf | Incumbent    BestBd   Gap | It/Node Time
...
Time limit reached
Best objective 1.245629365079e+02, best bound 1.241989911551e+02, gap 0.2922%

In [264]:
print("F: ", Int(sum(y_lr[:,1:38,4])/2), "\nSa: ", Int(sum(y_lr[:,1:38,5])/2), "\nSu: ", Int(sum(y_lr[:,1:38,6])/2), "\nM: ", Int(sum(y_lr[:,1:38,7])/2))

F: 39
Sa: 145
Su: 149
M: 47

In [265]:
CL = Int.(sum(u[i,1:38] for i in 1:20))
F = Int.(sum(y_lr[i,:,4] for i in 1:20)[1:38]/2)
SA = Int.(sum(y_lr[i,:,5] for i in 1:20)[1:38]/2)
SU = Int.(sum(y_lr[i,:,6] for i in 1:20)[1:38]/2)
M = Int.(sum(y_lr[i,:,7] for i in 1:20)[1:38]/2)

game_schedule = [CL F SA SU M]
#convert to csv
CSV.write("game_schedule.csv", DataFrame(game_schedule, ["Tue-Thu", "Fri", "Sat", "Sun", "Mon"]))


"game_schedule.csv"

In [267]:
for w in 1:38
    print(Int(sum(u[:,w])), "|")
end
println()
for w in 1:38
    print(Int(sum(y_lr[:,w,4])/2), "|")
end
println()
for w in 1:38
    print(Int(sum(y_lr[:,w,5])/2), "|")
end
println()
for w in 1:38
    print(Int(sum(y_lr[:,w,6])/2), "|")
end
println()
for w in 1:38
    print(Int(sum(y_lr[:,w,7])/2), "|")
end

0|0|0|0|6|6|6|0|0|0|0|6|6|6|0|0|0|0|0|0|0|0|0|0|0|0|0|0|0|0|0|0|0|0|0|0|0|0|
3|1|1|1|1|1|1|1|1|1|1|1|1|0|1|1|1|1|1|1|1|1|1|1|1|1|1|1|1|1|1|1|1|1|1|1|1|1|


4|4|4|4|4|3|2|4|4|4|4|3|3|3|4|4|4|4|4|4|4|4|4|4|4|4|4|4|4|4|4|4|4|4|4|4|4|3|
3|4|4|4|4|4|3|4|4|4|4|4|4|3|4|4|4|4|4|4|4|4|4|4|4|4|4|4|4|4|4|4|4|4|4|4|4|4|


0|1|1|1|1|2|4|1|1|1|1|2|2|4|1|1|1|1|1|1|1|1|1|1|1|1|1|1|1|1|1|1|1|1|1|1|1|2|

In [269]:
t1_schedule = zeros(38,7)
t1_schedule[:,4:7] = y_lr[1,1:38,:]
for w in 1:38
    for d in 4:7
        if t1_schedule[w,d] == 1
            t1_schedule[w,d] *= argmax(x[1,:,w]+x[:,1,w])
            if sum(x[:,1,w]) == 1
                t1_schedule[w,d] *= -1
            end
        end
    end
end
# t1_schedule[:,4:7] = Int.(t1_schedule[:,4:7])
# t1_schedule[1,1] = "*"
# t1_schedule

#create dataframe of strings
t1_schedule = DataFrame(Int.(t1_schedule), ["Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday", "Monday"])
for w in 1:38
    for d in 1:3
        if u[1,w] == 1 
            if z[1,w] == d
                t1_schedule[w,d] = 50
            end
        end
    end
end
t1_schedule
#convert to csv
CSV.write("t1_schedule.csv", t1_schedule)

"t1_schedule.csv"

In [15]:
function combine_both_models(u,z)
    N=20
    W=38
    D=7

    # create model
    model = Model(() -> Gurobi.Optimizer(GRB_ENV))
    set_optimizer_attribute(model, "TimeLimit", 300)
    set_optimizer_attribute(model, "OutputFlag", 1)
    # set_optimizer_attribute(model, "MIPGap", 1e-4)

    # VARIABLES
    @variable(model, y[i = 1:N, w = 1:W+1, d = 4:D], Bin) # 1 if team i plays on day d of week w, 0 otherwise
    #filler week W+1 to help with model
    # VARIABLES TEAM MATCH
    @variable(model, x[i = 1:N, j = 1:N, w = 1:W], Bin) # 1 if team i plays team j on week w, 0 otherwise

    # OBJECTIVE FUNCTION -- rest of one team doesn't depend on the other team, put that in the constraints
    @objective(model, Min, 
       sum( ( (1-u[i,w+1])*GetC(d,7+d_) + u[i,w+1]*(GetC(d,7+z[i,w+1])+GetC(z[i,w+1], d_)) )
           * (y[i,w,d]*y[i,w+1,d_])
       for i in 1:N, w in 1:W, d in 4:D, d_ in 4:D))
    # @objective(model, Min, 
    #         sum( ( C[7 * w + d, 7 * (w+1) + d_] )
    #             * (y[i,w,d]*y[i,w+1,d_])
    #         for i in 1:N, w in 1:W, d in 4:D, d_ in 4:D))

    # CONSTRAINTS
    
    # teams who play each other on week w must play on the same day of the week
    for w=1:W
        for i=1:N
            for j=1:N
                if x[i,j,w]+x[j,i,w] == 1
                    @constraint(model, (y[i,w,:] .== y[j,w,:]))
                end
            end
        end
    end

    # @constraint(model, [i = 1:N, j = 1:N, w = 1:W, d = 4:D], y[i,w,d] >= (x[i,j,w]+x[j,i,w])*y[j,w,d])

    # filler week 39
    @constraint(model, [i = 1:N], y[i,W+1,7] == 1)
    
    #team can only play as many times as scheduled
    @constraint(model, [i = 1:N, w = 1:W], sum(y[i,w,:]) == sum(x[i,:,w]+x[:,i,w]))

    # # maximum of 1 game on day 1
    # @constraint(model, [w = 1:W], sum(y[:,w,4]) <= 8) #2 teams play per game

    # # maximum of 4 games on day 2
    # @constraint(model, [w = 1:W], sum(y[:,w,5]) <= 12)

    # # maximum of 4 games on day 3
    # @constraint(model, [w = 1:W], sum(y[:,w,6]) <= 12)

    # # maximum of 1 game on day 4
    # @constraint(model, [w = 1:W], sum(y[:,w,7]) <= 8)

    # # at least 15 weekend home games
    # @constraint(model, [i = 1:N], sum(y[i,:,5]) + sum(y[i,:,6]) >= 15)
 
    # CONSTRAINTS
    # teams can never play against themselves
    @constraint(model, [i = 1:N, w = 1:W], x[i, i, w] == 0)

    # each team can only play at most one game per week
    @constraint(model, [i = 1:N, w = 1:W], sum(x[i, j, w] + x[j, i, w] for j in 1:N) <= 1)

    for i=1:N
        for j=1:N
            if i!=j
                # each game is against a different team, team play against each other team at most once in first half of season
                @constraint(model, sum(x[i, j, w] + x[j, i, w] for w in 1:Int(W/2)) == 1)
                
                # each game is against a different team, team play against each other team at most once in second half of season
                @constraint(model, sum(x[i, j, w] + x[j, i, w] for w in Int((W/2))+1:W) == 1)

                # every team plays every other team once at home
                @constraint(model, sum(x[i, j, w] for w in 1:W) == 1)
                
            end
        end
    end

    # no more than 2 straight home games
    @constraint(model, [i = 1:N, w = 1:W-2], sum(x[i, :, w] + x[i, :, w+1] + x[i, :, w+2]) <= 2)

    # no more than 2 straight away games
    @constraint(model, [j = 1:N, w = 1:W-2], sum(x[:, j, w] + x[:, j, w+1] + x[:, j, w+2]) <= 2)

    # OPTIMIZE
    # solvetime = @elapsed optimize!(model)
    optimize!(model)


    return value.(x), value.(y)
end

combine_both_models (generic function with 1 method)

In [16]:
x_comb,y_comb = combine_both_models(u,z)

Set parameter OutputFlag to value 1
Set parameter OutputFlag to value 1
Gurobi Optimizer version 9.5.2 build v9.5.2rc0 (mac64[arm])
Thread count: 8 physical cores, 8 logical processors, using up to 8 threads
Optimize a model with 4880 rows, 18320 columns and 192820 nonzeros
Model fingerprint: 0xde4dcc2f
Model has 12160 quadratic objective terms
Variable types: 0 continuous, 18320 integer (18320 binary)
Coefficient statistics:
  Matrix range     [1e+00, 2e+00]
  Objective range  [0e+00, 0e+00]
  QObjective range [2e-01, 2e+01]
  Bounds range     [0e+00, 0e+00]
  RHS range        [1e+00, 2e+00]
Presolve removed 1160 rows and 840 columns
Presolve time: 0.12s
Presolved: 15560 rows, 29320 columns, 207280 nonzeros
Variable types: 0 continuous, 29320 integer (29320 binary)

Deterministic concurrent LP optimizer: primal and dual simplex
Showing first log only...


Use crossover to convert LP symmetric solution to basic solution...
Concurrent spin time: 0.00s

Solved with dual simplex

Root rel

MathOptInterface.ResultIndexBoundsError{MathOptInterface.VariablePrimal}: Result index of attribute MathOptInterface.VariablePrimal(1) out of bounds. There are currently 0 solution(s) in the model.

In [33]:
function schedule_days_test(x,u,z)
    N=20
    W=38
    D=7 #playable number of days

    # create model
    model = Model(() -> Gurobi.Optimizer(GRB_ENV))
    set_optimizer_attribute(model, "TimeLimit", 7200)
    set_optimizer_attribute(model, "OutputFlag", 1)
    set_optimizer_attribute(model, "MIPGap", 1e-4)

    # VARIABLES
    @variable(model, y[i = 1:N, w = 1:W+1, d = 4:D], Bin) # 1 if team i plays on day d of week w, 0 otherwise
    @variable(model, a[i = 1:N, w = 1:W+1, d = 4:D, d_ = 4:D], Bin)
    #filler week W+1 to help with model

    # OBJECTIVE FUNCTION -- rest of one team doesn't depend on the other team, put that in the constraints
    @objective(model, Min, 
        sum( ( (1-u[i,w+1])*GetC(d,7+d_) + u[i,w+1]*(GetC(d,7+z[i,w+1])+GetC(z[i,w+1], d_)) )
           * (a[i,w,d,d_]) 
        for i in 1:N, w in 1:W, d in 4:D, d_ in 4:D))
    # @objective(model, Min, 
    #         sum( ( C[7 * w + d, 7 * (w+1) + d_] )
    #             * (y[i,w,d]*y[i,w+1,d_])
    #         for i in 1:N, w in 1:W, d in 4:D, d_ in 4:D))

    # # OBJECTIVE FUNCTION
    # @objective(model, Min, 
    #     sum( ( (1-u[i,w+1])*GetC(d,7+d_) + u[i,w+1]*(GetC(d,7+z[i,w+1])+GetC(z[i,w+1], d_)) )

    #         * (
    #             (x[i,j,w]+x[j,i,w])*(x[i,k,w+1]+x[k,i,w+1])
    #         * y[i,w,d]*y[i,w+1,d_]
    #         )

    #     for i in 1:N, j in 1:N, k in 1:N, w in 1:W, d in 4:D, d_ in 4:D))

    # CONSTRAINTS
    @constraint(model, [i = 1:N, w = 1:W, d = 4:D, d_ = 4:D], a[i,w,d,d_] >= y[i,w,d] + y[i,w+1,d_] - 1)
    @constraint(model, [i = 1:N, w = 1:W, d = 4:D, d_ = 4:D], a[i,w,d,d_] <= y[i,w,d])
    @constraint(model, [i = 1:N, w = 1:W, d = 4:D, d_ = 4:D], a[i,w,d,d_] <= y[i,w+1,d_])
    
    # teams who play each other on week w must play on the same day of the week
    for w=1:W
        for i=1:N
            for j=1:N
                if x[i,j,w]+x[j,i,w] == 1
                    @constraint(model, (y[i,w,:] .== y[j,w,:]))
                end
            end
        end
    end

    # filler week 39
    @constraint(model, [i = 1:N], y[i,W+1,7] == 1)
    
    #team can only play as many times as scheduled
    @constraint(model, [i = 1:N, w = 1:W], sum(y[i,w,:]) == sum(x[i,:,w]+x[:,i,w]))

    # maximum of 1 game on day 1
    @constraint(model, [w = 1:W], sum(y[:,w,4]) <= 2) #2 teams play per game

    # maximum of 4 games on day 2
    @constraint(model, [w = 1:W], sum(y[:,w,5]) <= 10)

    # maximum of 4 games on day 3
    @constraint(model, [w = 1:W], sum(y[:,w,6]) <= 10)

    # maximum of 1 game on day 4
    @constraint(model, [w = 1:W], sum(y[:,w,7]) <= 2)

    # # minimum of 3 games on day 2
    # @constraint(model, [w = 1:W], sum(y[:,w,5]) >= 6)

    # # minimum of 2 games on day 3
    # @constraint(model, [w = 1:W], sum(y[:,w,6]) >= 4)

    # at least 15 weekend home games
    @constraint(model, [i = 1:N], sum(y[i,:,5]) + sum(y[i,:,6]) >= 12)

    # # at least 1 Fri home game
    # @constraint(model, [i = 1:N], sum(y[i,:,4]) >= 1)

    # # at least 1 Mon home game
    # @constraint(model, [i = 1:N], sum(y[i,:,7]) >= 1)


    # OPTIMIZE
    # solvetime = @elapsed optimize!(model)
    optimize!(model)


    return value.(y)
end

function schedule_days_lr_test(x,u,z)
    N=20
    W=38
    D=7 #playable number of days

    # create model
    model = Model(() -> Gurobi.Optimizer(GRB_ENV))
    set_optimizer_attribute(model, "TimeLimit", 300)
    set_optimizer_attribute(model, "OutputFlag", 1)
    set_optimizer_attribute(model, "MIPGap", 1e-4)

    # VARIABLES
    @variable(model, y[i = 1:N, w = 1:W+1, d = 4:D], Bin) # 1 if team i plays on day d of week w, 0 otherwise
    @variable(model, a[i = 1:N, w = 1:W, d = 4:D, d_ = 4:D], Bin) # 1 if team i plays on day d of week w and day d_ of week w+1, 0 otherwise
    @variable(model, abs_f[w = 1:W, z=1:2] >= 0, Int)
    @variable(model, abs_sa[w = 1:W, z=1:2] >= 0, Int)
    @variable(model, abs_su[w = 1:W, z=1:2] >= 0, Int)
    @variable(model, abs_m[w = 1:W, z=1:2] >= 0, Int)
    #@variable(model, P) 
    #@variable(model, lambda[i = 1:4, z=1:2]>=0)
    #filler week W+1 to help with model

    # the bigger the value, the tighter the constraint
    l_1_day = 0.015          #punish from exactness
    l_2_day = 0.01          #punish for breaking range
    l_1_end = 0.005          #punish for exactness
    l_2_end = 0.01          #punish for breaking range

    # OBJECTIVE FUNCTION -- rest of one team doesn't depend on the other team, put that in the constraints
    @objective(model, Min, sum( ((1-u[i,w+1])*GetC(d,7+d_) + u[i,w+1]*(GetC(d,7+z[i,w+1])+GetC(z[i,w+1], d_))) * (a[i,w,d,d_]) for i in 1:N, w in 1:W, d in 4:D, d_ in 4:D)
        + sum(l_1_day*(abs_f[w,1]) + l_1_end*(abs_sa[w,1]) + l_1_end*(abs_su[w,1]) + l_1_day*(abs_m[w,1]) 
        + l_2_day*(abs_f[w,2]) + l_2_end*(abs_sa[w,2]) + l_2_end*(abs_su[w,2]) + l_2_day*(abs_m[w,2]) for w=1:W)
        )

    # @objective(model, Min, 
    #         sum( ( C[7 * w + d, 7 * (w+1) + d_] )
    #             * (y[i,w,d]*y[i,w+1,d_])
    #         for i in 1:N, w in 1:W, d in 4:D, d_ in 4:D))

    # CONSTRAINTS
    @constraint(model, [i = 1:N, w = 1:W, d = 4:D, d_ = 4:D], a[i,w,d,d_] >= y[i,w,d] + y[i,w+1,d_] - 1)
    @constraint(model, [i = 1:N, w = 1:W, d = 4:D, d_ = 4:D], a[i,w,d,d_] <= y[i,w,d])
    @constraint(model, [i = 1:N, w = 1:W, d = 4:D, d_ = 4:D], a[i,w,d,d_] <= y[i,w+1,d_])

    # punish schedule for deviating from 1, 4, 4, 1 schedule
    # punish schedule even more for deviating from 0-2, 3-5, 3-5, 0-2 schedule
    @constraint(model, [w = 1:W], abs_f[w,2] >= (sum(y[:,w,4]) - 6))
    @constraint(model, [w = 1:W], abs_f[w,1] >= (sum(y[:,w,4]) - 2))
    @constraint(model, [w = 1:W], abs_f[w,1] >= (2 - sum(y[:,w,4])))

    @constraint(model, [w = 1:W], abs_sa[w,2] >= (sum(y[:,w,5]) - 10))
    @constraint(model, [w = 1:W], abs_sa[w,2] >= (6 - sum(y[:,w,5])))
    @constraint(model, [w = 1:W], abs_sa[w,1] >= (sum(y[:,w,5]) - 8))
    @constraint(model, [w = 1:W], abs_sa[w,1] >= (8 - sum(y[:,w,5])))

    @constraint(model, [w = 1:W], abs_su[w,2] >= (sum(y[:,w,6]) - 10))
    @constraint(model, [w = 1:W], abs_su[w,2] >= (6 - sum(y[:,w,6])))
    @constraint(model, [w = 1:W], abs_su[w,1] >= (sum(y[:,w,6]) - 8))
    @constraint(model, [w = 1:W], abs_su[w,1] >= (8 - sum(y[:,w,6])))

    @constraint(model, [w = 1:W], abs_m[w,2] >= (sum(y[:,w,7]) - 6))
    @constraint(model, [w = 1:W], abs_m[w,1] >= (sum(y[:,w,4]) - 4))
    @constraint(model, [w = 1:W], abs_m[w,1] >= (2 - sum(y[:,w,7])))
   


    # teams who play each other on week w must play on the same day of the week
    for w=1:W
        for i=1:N
            for j=1:N
                if x[i,j,w]+x[j,i,w] == 1
                    @constraint(model, (y[i,w,:] .== y[j,w,:]))
                end
            end
        end
    end

    # @constraint(model, [i = 1:N, j = 1:N, w = 1:W, d = 4:D], y[i,w,d] >= (x[i,j,w]+x[j,i,w])*y[j,w,d])

    # filler week 39
    @constraint(model, [i = 1:N], y[i,W+1,7] == 1)
    
    #team can only play as many times as scheduled
    @constraint(model, [i = 1:N, w = 1:W], sum(y[i,w,:]) == sum(x[i,:,w]+x[:,i,w]))

    # # maximum of 6 game on day 2
    # @constraint(model, [w = 1:W], sum(y[:,w,5]) <= 12) #2 teams play per game

    # # maximum of 6 game on day 3
    # @constraint(model, [w = 1:W], sum(y[:,w,6]) <= 12) #2 teams play per game

    # # minimum of 3 game on day 2
    # @constraint(model, [w = 1:W], sum(y[:,w,5]) >= 6) #2 teams play per game

    # # minimum of 3 game on day 3
    # @constraint(model, [w = 1:W], sum(y[:,w,6]) >= 6) #2 teams play per game

    # at least 12 weekend home games
    @constraint(model, [i = 1:N], sum(y[i,:,5]) + sum(y[i,:,6]) >= 12)


    # OPTIMIZE
    # solvetime = @elapsed optimize!(model)
    optimize!(model)


    return value.(y)
end

schedule_days_lr_test (generic function with 1 method)

In [35]:
y_test = schedule_days_test(x,u,z)

Set parameter OutputFlag to value 1
Set parameter OutputFlag to value 1
Gurobi Optimizer version 9.5.2 build v9.5.2rc0 (mac64[arm])
Thread count: 8 physical cores, 8 logical processors, using up to 8 threads
Optimize a model with 40472 rows, 15600 columns and 98860 nonzeros
Model fingerprint: 0xfa5d005f
Variable types: 0 continuous, 15600 integer (15600 binary)
Coefficient statistics:
  Matrix range     [1e+00, 1e+00]
  Objective range  [1e-01, 1e+01]
  Bounds range     [0e+00, 0e+00]
  RHS range        [1e+00, 1e+01]
Presolve removed 27840 rows and 1940 columns
Presolve time: 0.14s
Presolved: 12632 rows, 13660 columns, 40840 nonzeros
Variable types: 0 continuous, 13660 integer (13660 binary)

Root relaxation: objective 2.311111e+00, 1975 iterations, 0.05 seconds (0.06 work units)

    Nodes    |    Current Node    |     Objective Bounds      |     Work
 Expl Unexpl |  Obj  Depth IntInf | Incumbent    BestBd   Gap | It/Node Time

     0     0    2.31111    0  764          -    2.31111 

3-dimensional DenseAxisArray{Float64,3,...} with index sets:
    Dimension 1, Base.OneTo(20)
    Dimension 2, Base.OneTo(39)
    Dimension 3, 4:7
And data, a 20×39×4 Array{Float64, 3}:
[:, :, 4] =
 -0.0  -0.0  -0.0   1.0   0.0  -0.0  …  -0.0   0.0  -0.0  -0.0  -0.0  0.0
 -0.0  -0.0  -0.0  -0.0   0.0  -0.0     -0.0  -0.0   0.0  -0.0  -0.0  0.0
 -0.0  -0.0  -0.0   0.0   0.0   0.0     -0.0   0.0  -0.0  -0.0  -0.0  0.0
 -0.0  -0.0  -0.0   0.0   0.0   0.0     -0.0   0.0   0.0   0.0  -0.0  0.0
 -0.0  -0.0  -0.0  -0.0   0.0   0.0      0.0   0.0  -0.0   0.0   0.0  0.0
  0.0  -0.0  -0.0   0.0   0.0   0.0  …   0.0   0.0  -0.0   0.0  -0.0  0.0
  1.0   0.0   0.0  -0.0   0.0  -0.0      0.0   0.0  -0.0   0.0  -0.0  0.0
  0.0   0.0  -0.0  -0.0   0.0  -0.0      0.0  -0.0   0.0   0.0  -0.0  0.0
 -0.0  -0.0  -0.0   0.0  -0.0   0.0     -0.0   0.0   0.0   0.0  -0.0  0.0
  0.0   0.0  -0.0   0.0   0.0   0.0     -0.0   0.0   0.0  -0.0   0.0  0.0
  0.0   0.0  -0.0   0.0   0.0  -0.0  …   0.0  -0.0   0.0   0.0 

120 MINUTES, linearized HARD DAY CONSTRAINTS

Set parameter OutputFlag to value 1
Set parameter OutputFlag to value 1
Gurobi Optimizer version 9.5.2 build v9.5.2rc0 (mac64[arm])
Thread count: 8 physical cores, 8 logical processors, using up to 8 threads
Optimize a model with 40472 rows, 15600 columns and 98860 nonzeros
Model fingerprint: 0xfa5d005f
Variable types: 0 continuous, 15600 integer (15600 binary)
Coefficient statistics:
  Matrix range     [1e+00, 1e+00]
  Objective range  [1e-01, 1e+01]
  Bounds range     [0e+00, 0e+00]
  RHS range        [1e+00, 1e+01]
Presolve removed 27840 rows and 1940 columns
Presolve time: 0.12s
Presolved: 12632 rows, 13660 columns, 40840 nonzeros
Variable types: 0 continuous, 13660 integer (13660 binary)

Root relaxation: objective 2.311111e+00, 1975 iterations, 0.05 seconds (0.06 work units)

    Nodes    |    Current Node    |     Objective Bounds      |     Work
 Expl Unexpl |  Obj  Depth IntInf | Incumbent    BestBd   Gap | It/Node Time

     0     0    2.31111    0  764          -    2.31111      -     -    0s
H    0     0                     132.1460317    2.31111  98.3%     -    0s
H    0     0                     125.5507937    2.31111  98.2%     -    0s
...
Time limit reached
Best objective 1.246523809524e+02, best bound 1.245092817979e+02, gap 0.1148%